torch_geometric/explain/explainer.py

```diff
@@ -0,0 +1,342 @@
+from dataclasses import dataclass
+from inspect import getfullargspec
+from typing import Optional, Union
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.explain.base import ExplainerAlgorithm
+from torch_geometric.explain.explanations import Explanation
+
+
+@dataclass
+class Threshold:
+    """Class to store and validate threshold parameters."""
+    type: str
+    value: Union[float, int]
+    _valid_type = ["none", "hard", "topk", "connected", "topk_hard"]
+
+    def __post_init__(self):
+        if self.type not in self._valid_type:
+            raise ValueError(f'Invalid threshold type {self.type}. '
+                             f'Valid types are {self._valid_type}.')
+        if not isinstance(self.value, (float, int)):
+            raise ValueError(f'Invalid threshold value {self.value}. '
+                             f'Valid values are float or int.')
+
+        if self.type == "hard":
+            if self.value < 0 or self.value > 1:
+                raise ValueError(f'Invalid threshold value {self.value}. '
+                                 f'Valid values are in [0, 1].')
+        if self.type in ["topk", "connected", "topk_hard"]:
+            if self.value <= 0 or not isinstance(self.value, int):
+                raise ValueError(f'Invalid threshold value {self.value}. '
+                                 f'Valid values are positif integers.')
+
+
+@dataclass
+class ExplainerConfig:
+    """Class to store and validate high level explanation parameters."""
+    explanation_type: str
+    mask_type: str
+    _valid_explanation_type = ["model", "phenomenon"]
+    _valid_mask_type = [
+        "node", "edge", "node_and_edge", "layers", "node_feat", "edge_feat",
+        "node_feat_and_edge_feat", "node_feat_and_edge", "node_and_edge_feat"
```

Comment:
Wonder if we can separate the `mask_type` into `node_mask_type` and `edge_mask_type` for a cleaner interface.